### PR TITLE
Add yog code

### LINKED ISSUE #4: Changes to phys routines
### Issue Type

Other (please describe below)

### Issue Description

Bring changes to phys routines over from CAM-ML development into the tag branch.

As detailed on https://github.com/m2lines/convection-parameterization-in-CAM/tree/cam-interface/CAM_mods

### Will this change answers?

Yes

### Will you be implementing this yourself?

Yes

### PR DESCRIPTION
**Pull request by jatkinson1000**
This adds calls to the YOG routine in physpkg in CAM, whilst also running ZM.
Should close #4 

- A call to initialise the net is made in `phys_init` - https://github.com/m2lines/CAM-ML/pull/9/commits/19886bf8a8f824fdb2752bb17f19a590b5489cc6
- A call to the net is made ~before~ after `convect_deep_tend` in `tphysbc` - https://github.com/m2lines/CAM-ML/pull/9/commits/dc0a005b229075deb3cb3139a721e9a5e35197d2 and https://github.com/m2lines/CAM-ML/pull/9/commits/5df40aa7ba38e3b3ddc6514f204668a705d7e1e6
    - This also includes timing of the YOG scheme using the inbuilt CESM timing infrastructure
- Cleanup of the net and weights is made in https://github.com/m2lines/CAM-ML/pull/9/commits/ee3005be9331e9705422db3e86c359fe8a62cdc2

Next steps:

- [ ] Take the multiple call blocks into a more contained routines:
    - Net init and addfld into a yog_init routine
    - ptend_init, net call, outfld into a yog_main routine
- [ ] Add the NN code currently held in SourceMods.
- [ ] Update the xml as discussed with @paogorman to have a separate yog parameter, rather than assigning as an option to `deep_scheme`. This will be done in a refactor of #2 which this should then be rebased off of.

### CODE REVIEW

**Comment by jatkinson1000** [comment]
As discussed with @paogorman we would prefer to call the YOG scheme **after** deep convection.
This is applied in https://github.com/m2lines/CAM-ML/pull/9/commits/5df40aa7ba38e3b3ddc6514f204668a705d7e1e6

**Review by casterkay — APPROVED** [review]
LGTM

**Comment by MarionBWeinzierl** [comment]
> This adds calls to the YOG routine in physpkg in CAM, whilst also running ZM. Should close #4
> 
>     * A call to initialise the net is made in `phys_init` - [19886bf](https://github.com/m2lines/CAM-ML/commit/19886bf8a8f824fdb2752bb17f19a590b5489cc6)
> 
>     * A call to the net is made ~before~ after `convect_deep_tend` in `tphysbc` - [dc0a005](https://github.com/m2lines/CAM-ML/commit/dc0a005b229075deb3cb3139a721e9a5e35197d2) and [5df40aa](https://github.com/m2lines/CAM-ML/commit/5df40aa7ba38e3b3ddc6514f204668a705d7e1e6)
>       
>       * This also includes timing of the YOG scheme using the inbuilt CESM timing infrastructure
> 
>     * Cleanup of the net and weights is made in [ee3005b](https://github.com/m2lines/CAM-ML/commit/ee3005be9331e9705422db3e86c359fe8a62cdc2)
> 
> 
> Next steps:
> 
>     * [ ]  Take the multiple call blocks into a more contained routines:
>       
>       * Net init and addfld into a yog_init routine
>       * ptend_init, net call, outfld into a yog_main routine
> 
>     * [ ]  Add the NN code currently held in SourceMods.
> 
>     * [ ]  Update the xml as discussed with @paogorman to have a separate yog parameter, rather than assigning as an option to `deep_scheme`. This will be done in a refactor of [Xml updates for YOG #2](https://github.com/m2lines/CAM-ML/pull/2) which this should then be rebased off of.

Just to clarify, are the "next steps" something you are doing in this PR, or a new one? As it's a checklist for this one (which is not checked). If it's done in a different PR, it would be good to move the list there (or into the associated issue, that's a matter of preference), and link that from here.

File: bld/namelist_files/namelist_definition.xml [hunk 1]
```diff
@@ -2955,12 +2955,31 @@ Default: 10
...
 Type of deep convection scheme employed.  'ZM' for Zhang-McFarlane;
-'off' for none; or 'UNICON' which doesn't distinquish shallow and deep.
+'off' for none;'YOG' for Yuval-O'Gorman; or 'UNICON' which doesn't
+distinquish shallow and deep.
 Default: 'ZM' unless using 'UNICON', 'SPCAM' or 'pbl=none'
 </entry>
 
+<entry id="run_deep_comp" type="char*16" category="conv"
+       group="phys_ctl_nl" valid_values="off,on">
+Whether or not to use both schemes for deep convection.
```
Comment (MarionBWeinzierl) [inline bld/namelist_files/namelist_definition.xml:2967]: See my comment here: https://github.com/m2lines/CAM-ML/pull/2#discussion_r1560685630

File: src/physics/cam/convect_deep.F90 [hunk 2]
```diff
@@ -273,6 +279,17 @@ subroutine convect_deep_tend( &
           jctop, jcbot , &
           state   ,ptend   ,landfrac, pbuf)
 
+  case('YOG') !    Yuval-O'Gorman (running ZM for now)
```
Comment (MarionBWeinzierl) [inline src/physics/cam/convect_deep.F90:282]: Is this not running YOG yet? If not, a warning should be printed

File: src/physics/cam/convect_deep.F90 [hunk 3]
```diff
@@ -307,7 +324,10 @@ subroutine convect_deep_tend_2( state,  ptend,  ztodt, pbuf)
 
    if ( deep_scheme .eq. 'ZM' ) then  ! Zhang-McFarlane
       call zm_conv_tend_2( state,   ptend,  ztodt,  pbuf) 
-   else
+   else if ( deep_scheme .eq. 'YOG' ) then  ! Yuval-O'Gorman
+      ! TODO Do nothing extra for now as seems to be a separate scheme
```
Comment (MarionBWeinzierl) [inline src/physics/cam/convect_deep.F90:328]: I don't understand this comment, could you explain?

File: src/physics/cam/physpkg.F90 [hunk 4]
```diff
@@ -1923,9 +1955,38 @@ subroutine tphysbc (ztodt, state,  &
          state,   ptend, cam_in%landfrac, pbuf)
 
     call physics_update(state, ptend, ztodt, tend)
+    ! NB: If we want to discard ZM tendencies here we need to call a ptend_dealloc
```
Comment (MarionBWeinzierl) [inline src/physics/cam/physpkg.F90:1942]: Does that need to be added? Is this a new issue that should be raised? And what are "ZM tendencies"?

File: src/physics/cam/physpkg.F90 [hunk 5]
```diff
@@ -1923,9 +1955,38 @@ subroutine tphysbc (ztodt, state,  &
...
+        call physics_ptend_init(ptend, state%psetcols, 'yogNN', ls=.true., lq=.true.)! initialize ptend type for YOG
+
+        call nn_convection_flux_CAM(state%pmid(:,pver:1:-1), state%pint(:,pverp:1:-1), state%ps, &
+                                    state%t(:,pver:1:-1), state%q(:,pver:1:-1,1), &
+                                    state%q(:,pver:1:-1,ixcldliq), state%q(:,pver:1:-1,ixcldice), &
+                                    cpair, &
+                                    ztodt, &
+                                    ncol, pver, &
+                                    1, 1, 1, &
+                                    dummy_prec, &
```
Comment (MarionBWeinzierl) [inline src/physics/cam/physpkg.F90]: What is "dummy_prec" for, and why do you need a dummy?

File: src/physics/cam/yog_conv_intr.F90 [hunk 6]
```diff
@@ -0,0 +1,923 @@
+module yog_conv_intr
```
Comment (MarionBWeinzierl) [inline src/physics/cam/yog_conv_intr.F90]: This is a big file with some very long subroutines. Can it be modularised/subdivided further, to make it easier to read and adapt?

**Review by MarionBWeinzierl — COMMENTED** [review]
I have a couple of questions and comments about this PR, see below.

Before approving I'd also like to confirm my question above about the "next steps".

**Comment by jatkinson1000** [comment]
As commented on in #14 this PR was was superseded by that PR so I am closing this now.

**Pull request closed without merging**